File: 22-dp1/3-tabulation/main.cpp
// Tabulation - C++ example

#include <iostream>
#include <vector>

int fib(int n) {
    if (n <= 1) return n;
    std::vector<int> dp(n+1, 0);
    dp[1] = 1;

    for (int i = 2; i <= n; ++i)
        dp[i] = dp[i-1] + dp[i-2];

    return dp[n];
}

int main() {
    std::cout << fib(10) << std::endl;  // Output: 55
    return 0;
}

File: 22-dp1/3-tabulation/main.py
# Tabulation - Python example

def fib(n):
    if n <= 1:
        return n
    dp = [0] * (n+1)
    dp = 1

    for i in range(2, n+1):
        dp[i] = dp[i-1] + dp[i-2]

    return dp[n]

if __name__ == "__main__":
    print(fib(10))  # Output: 55
